scenarios/smart_trailer/digital_twin_providers/trailer_properties_provider/src/trailer_properties_provider_impl.rs

```diff
@@ -170,15 +171,17 @@ impl TrailerPropertiesProviderImpl {
 
             for constraint in constraints {
                 if constraint.r#type == *FREQUENCY_MS {
-                    frequency_ms = u64::from_str(&constraint.value).unwrap();
+                    frequency_ms = u64::from_str(&constraint.value).map_err(|err| {
+                        format!("Failed to parse frequency constraint due to '{err:?}'")
+                    })?;
                 };
             }
 
             loop {
                 // See if we need to shutdown.
                 if reciever.try_recv() == Err(mpsc::error::TryRecvError::Disconnected) {
                     info!("Shutdown thread for {topic}.");
-                    return;
+                    return Result::<(), String>::Ok(());
```

Comment:
This syntax looks weird. I did some research and I think ultimately it's here because I suggested usage of `?` which caused a big cascade of side-effects. Using `?` requires you're in a block that returns `Result`. This code is in an async block though, and for some reason you can't specify the return type of an async block when you declare it, it's only determined implicitly. Then, since you're only ever returning `Ok(())` the compiler can't infer the error type without some help, and this is one way to give it that hint (though I'm surprised it didn't figure it out from line 174-176, which should be enough info to figure out it's `String`).

With all that said, I think this will be cleaner if you define the return type when you spawn the task, similar to what you did on line 123 of scenarios/smart_trailer/applications/smart_trailer_application/src/main.rs. Then you should be able to use `Ok(())` here and below. We could also go back to using `unwrap` since it won't take down the entire program (see https://doc.servo.org/tokio/runtime/task/join/struct.JoinHandle.html), though it may be harder to figure out what's happening in error cases and it won't be as easy if we want to add error handling to this later

Reply:
Yes, that is exactly why this was added. I will fix it to match the smart_trailer_application main.rs